### PR TITLE
🌱 test/e2e: Cleanup test files & update doc

### LINKED ISSUE #7699: update `test/e2e/data/infrastructure-docker/README.md` with more details on test files setup
**Detailed Description**

Update the readme with more details on :
- Why we need multiple version folders under a particular contract version, ex: `v1beta1`. 
- How these files should be maintained
- When a folder can be dropped.

**Anything else you would like to add:**

[Miscellaneous information that will assist in solving the issue.]

/kind documentation
/area testing

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
I know the diff looks a bit scary but mostly just moved or deleted files.

Changes:
* `data/infrastructure` and `data/shared` are now organized consistently by version (not partially by contract)
* Dropped v1.2 => main test as we are not supporting (and thus not testing) skip-level upgrades. Should also speed up our test jobs.
* Updated documentation accordingly

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7699
Follow-up to https://github.com/kubernetes-sigs/cluster-api/pull/7692#discussion_r1055474785

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,11 +11,10 @@ hack/tools/bin
 *.test
```
Comment (sbueringer) [inline .gitignore:12]: Note to myself:
* [ ] Improve corresponding docs (test/e2e/config/README.md + release tasks doc)

**Comment by sbueringer** [comment]
/assign @fabriziopandini @killianmuldoon @ykakarap 

Please see changes documented in the PR description

File: docs/release/release-tasks.md [hunk 2]
```diff
@@ -90,11 +90,20 @@ The goal of this issue is to bump the versions on the main branch so that the up
 is used for e.g. local development and e2e tests. We also modify tests so that they are testing the previous release.
 
 This comes down to changing occurrences of the old version to the new version, e.g. `v1.3` to `v1.4`:
-1. Add new release to top-level `metadata.yaml` and `test/e2e/data/shared/v1beta1/metadata.yaml`.
+1. Create a new `v1.3` `metadata.yaml` (`test/e2e/data/shared/v1.3/metadata.yaml`) by copying
+   `test/e2e/data/shared/main/metadata.yaml`
+1. Add the new release to the main `metadata.yaml` (`test/e2e/data/shared/main/metadata.yaml`).
```
Comment (fabriziopandini) [inline docs/release/release-tasks.md:95]: What about grouping items related to E2E under a single bullet? Something like
   
```suggestion
1. Add the new release to the main `metadata.yaml` (`test/e2e/data/shared/main/metadata.yaml`).
2. Setup E2E tests for the new release
   3. Create a new `v1.3` `metadata.yaml` (`test/e2e/data/shared/v1.3/metadata.yaml`) by copying
   `test/e2e/data/shared/main/metadata.yaml`
   .... (everything till L106)
   5. Make sure all tests are green (also run `pull-cluster-api-e2e-full-main` and `pull-cluster-api-e2e-workload-upgrade-1-23-latest-main`).
3. Update `create-local-repository.py` and `tools/tilt-prepare/main.go` with the new release: `v1.3.99` => `v1.4.99`.
```

File: test/e2e/clusterctl_upgrade_test.go [hunk 3]
```diff
@@ -58,21 +58,6 @@ var _ = Describe("When testing clusterctl upgrades (v0.4=>current)", func() {
 	})
 })
 
-var _ = Describe("When testing clusterctl upgrades (v1.2=>current)", func() {
```
Comment (fabriziopandini) [inline test/e2e/clusterctl_upgrade_test.go:61]: NOTE: I think we should document somewhere that step upgrades are not supported

File: test/e2e/data/infrastructure-docker/README.md [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 
 This folder contains cluster templates used in e2e tests. Each sub-folder contains cluster templates
-complying to the corresponding contract.
+for the corresponding Cluster API version.
 
-The templates in [v1alpha3](./v1alpha3), [v1alpha4](./v1alpha4) and [v1.2](./v1beta1/v1.2) are used for 
-clusterctl upgrade tests. In those tests we first deploy an old version of Cluster API and then upgrade 
-to the version from the current branch. We can't use the same templates for `v1.2` and `main` as not all 
-fields we use in `main` already existed in `v1.2`.
+The cluster templates are used for clusterctl upgrade tests. In those tests we first deploy an old version of Cluster API and then upgrade to the version from the current branch.
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/README.md:6]:
```suggestion
Subfolders for the old versions of Cluster API should contain only templates clusterctl upgrade tests; in those tests we first deploy an old version of Cluster API, create a workload cluster and then upgrade Cluster API to the version from the current branch (and check that the workload cluster works fine).
```

**Review by fabriziopandini — COMMENTED** [review]
I really like the new organizations it is cleaner and simpler to maintain. Great job

File: test/e2e/clusterctl_upgrade_test.go (hunk 3 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:61]: Absolutely fine for me. Do you have a suggestion where? Is it okay to do this in a follow-up PR given that I think it's unrelated to this PR.

I think it's debatable what sort of upgrades Cluster API / clusterctl supports. I think clusterctl allows a lot, especially when the contract is the same (not sure). But up until now nobody actually ran e2e tests for skip-level upgrades and I wouldn't want to put the burden on the community / us to start supporting it (that's why I'm dropping those tests in this PR, we didn't have them for older CAPI releases)

File: docs/release/release-tasks.md (hunk 2 above)
Comment (sbueringer) [reply docs/release/release-tasks.md:95]: Good idea, done

File: test/e2e/clusterctl_upgrade_test.go (hunk 3 above)
Comment (fabriziopandini) [reply test/e2e/clusterctl_upgrade_test.go:61]: No need to do this in this PR.
I think that for starting we can document this in the clusterctl update section (and maybe make sure it enforces this rule)

**Review by fabriziopandini — COMMENTED** [review]
/lgtm

**Comment by sbueringer** [comment]
@ykakarap @killianmuldoon If you have some time would be good if you can take a look. It's mostly just moving directories around. I woudl like to move ahead with this change (+ cherry-picks) relatively soon to avoid running into conflicts with other PRs.

File: test/e2e/clusterctl_upgrade_test.go (hunk 3 above)
Comment (sbueringer) [reply test/e2e/clusterctl_upgrade_test.go:61]: Discussed with Fabrizio. Given v1.2 is still in support (in previous releases we only had one v1beta1 version in support) we decided to re-add v1.2 tests.

**Comment by sbueringer** [comment]
@killianmuldoon just fyi, pushed another commit (because of https://github.com/kubernetes-sigs/cluster-api/pull/7825#discussion_r1062492100)

**Review by killianmuldoon — APPROVED** [review]
/lgtm 

The new organisation makes a lot of sense to me - thanks for doing this!

File: test/e2e/clusterctl_upgrade.go [hunk 5]
```diff
@@ -293,9 +293,9 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			input.E2EConfig.GetProviderLatestVersionsByContract(initContract, config.KubeadmControlPlaneProviderName))
 		infrastructureProviders = getValueOrFallback(input.InitWithInfrastructureProviders,
 			input.E2EConfig.GetProviderLatestVersionsByContract(initContract, input.E2EConfig.InfrastructureProviders()...))
-		ipamProviders = getValueOrFallback(input.IPAMProviders,
+		ipamProviders = getValueOrFallback(input.InitWithIPAMProviders,
 			input.E2EConfig.GetProviderLatestVersionsByContract(initContract, input.E2EConfig.IPAMProviders()...))
-		runtimeExtensionProviders = getValueOrFallback(input.RuntimeExtensionProviders,
+		runtimeExtensionProviders = getValueOrFallback(input.InitWithRuntimeExtensionProviders,
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:298]: That's a fun one :)

File: test/e2e/clusterctl_upgrade.go [hunk 6]
```diff
@@ -735,7 +735,7 @@ func matchUnstructuredLists(l1 *unstructured.UnstructuredList, l2 *unstructured.
 
 // getValueOrFallback returns the input value unless it is empty, then it returns the fallback input.
 func getValueOrFallback(value []string, fallback []string) []string {
-	if len(value) > 0 {
+	if value != nil {
```
Comment (sbueringer) [inline test/e2e/clusterctl_upgrade.go:738]: I think it's better to use value as soon as it's not nil. This allows folks to overwrite a value with an empty slice (as needed in our use case for runtime extension providers)

**Comment by sbueringer** [comment]
Found another bug in our clusterctl test framework code. Fix upcoming.. :)

File: test/framework/clusterctl/client.go [hunk 7]
```diff
@@ -96,24 +88,33 @@ func Init(_ context.Context, input InitInput) {
...
 	args := []string{"init", "--config", input.ClusterctlConfigPath, "--kubeconfig", input.KubeconfigPath}
 	if input.CoreProvider != "" {
 		args = append(args, "--core", input.CoreProvider)
 	}
 	if len(input.BootstrapProviders) > 0 {
 		args = append(args, "--bootstrap", strings.Join(input.BootstrapProviders, ","))
 	}
+	if len(input.ControlPlaneProviders) > 0 {
+		args = append(args, "--control-plane", strings.Join(input.ControlPlaneProviders, ","))
+	}
```
Comment (sbueringer) [inline test/framework/clusterctl/client.go:117]: Previously we did not set `--control-plane` in InitWithBinary.

The result for the v1.2 => main test was that the following providers where installed during init:
```
 k get provider -A
NAMESPACE                           NAME                    AGE   TYPE                     PROVIDER      VERSION
capd-system                         infrastructure-docker   8s    InfrastructureProvider   docker        v1.2.8
capi-kubeadm-bootstrap-system       bootstrap-kubeadm       9s    BootstrapProvider        kubeadm       v1.2.8
capi-kubeadm-control-plane-system   control-plane-kubeadm   8s    ControlPlaneProvider     kubeadm       v1.4.99
capi-system                         cluster-api             9s    CoreProvider             cluster-api   v1.2.8
```

Comment (sbueringer) [reply test/framework/clusterctl/client.go:117]: This only seems to be an issue if there are newer versions of the same contract, e.g. the v0.4 => main test was okay:
```
k get provider -A
NAMESPACE                           NAME                    AGE   TYPE                     PROVIDER      VERSION
capd-system                         infrastructure-docker   26s   InfrastructureProvider   docker        v0.4.8
capi-kubeadm-bootstrap-system       bootstrap-kubeadm       26s   BootstrapProvider        kubeadm       v0.4.8
capi-kubeadm-control-plane-system   control-plane-kubeadm   26s   ControlPlaneProvider     kubeadm       v0.4.8
capi-system                         cluster-api             26s   CoreProvider             cluster-api   v0.4.8
```

**Comment by sbueringer** [comment]
Squashed

**Comment by sbueringer** [comment]
(Just pushed a small fix as I broke the spacing in the clusterctl init log with strings.Join)

/test pull-cluster-api-e2e-full-main
/test pull-cluster-api-e2e-workload-upgrade-1-25-latest-main

**Comment by sbueringer** [comment]
/retest

(flake)

**Comment by fabriziopandini** [comment]
/hold 
for after the patch release

File: docs/release/release-tasks.md [hunk 8]
```diff
@@ -90,13 +90,28 @@ The goal of this issue is to bump the versions on the main branch so that the up
...
+      1. Create a new `v1.3` `metadata.yaml` (`test/e2e/data/shared/v1.3/metadata.yaml`) by copying
+   `test/e2e/data/shared/main/metadata.yaml`
+      2. Add the new release to the main `metadata.yaml` (`test/e2e/data/shared/main/metadata.yaml`).
+      3. Remove old `metadata.yaml`'s that are not used anymore (for `v1.4` we don't have to remove any).
+   4. Adjust cluster templates in `test/e2e/data/infrastructure-docker`:
+      1. Create a new `v1.3` folder. It should be created based on the `main` folder and only contain the templates
+         we use in the clusterctl upgrade tests (as of today `cluster-template` and `cluster-template-topology`).
+      2. Remove old folders that are not used anymore (for `v1.4` we don't have to remove any).
+   5. Modify the test specs in `test/e2e/clusterctl_upgrade_test.go` (according to the versions we want to test described above).
+      Please note that `InitWithKubernetesVersion` should be the highest mgmt cluster version supported by the respective Cluster API version.
```
Comment (sbueringer) [inline docs/release/release-tasks.md:112]: Rebased & added this line

**Comment by fabriziopandini** [comment]
/lgtm
/approve

@sbueringer feel free to leave the hold when you think we can merge this according to the other work going on on tests/test flakes

**Comment by sbueringer** [comment]
Will open the release-1.2 cherry-pick soon as well

**Pull request merged**